### DISCUSSION TITLE
[Question][Hw5] About the types of expressions

### DISCUSSION THREAD

**Discussion by kws1207** [prosyslab-classroom/cs348-information-security#592] [Q&A]
![image](https://user-images.githubusercontent.com/60031762/119862892-f6477100-bf53-11eb-8857-e2bfe50c2bcc.png)

In semantics.ml, we have to write the function
`let eval e mem`
according to above definition.

However, I am having a hard time figuring out what is the type of arithmetic operation and comparison operation.
Is it an inherited type of Llvm.llvalue? or is it defined in domain.ml? I can't find it ..

**Comment by Standchen** [comment]
I believe that you've already implemented fuctions for arithmetic/comparison operators in abstract domains in `domain.ml`. When you utilize those functions in `eval`, then you'll get what you want (some values with type `Value.t`.)

**Comment by KihongHeo** [comment]
I don't fully understand your questions but let me give some comments

- LLVM has a slightly different syntactic structure from the document. You can handle each arithmetic operator directly as an instruction. Please take a look at `Llvm.instr_opcode`
- Also `Llvm.classify_value` may be helpful.
- Take a look at this [document](https://github.com/prosyslab-classroom/llvm-primer) for general information.

**Comment by kws1207** [comment]
Thank you.
